network/vpack/msgp.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (C) 2019-2025 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package vpack
+
+import (
+	"fmt"
+)
+
+// Minimal msgpack constants used here
+const (
+	fixMapMask = 0x80
+	fixMapMax  = 0x8f
+	fixStrMask = 0xa0
+	fixStrMax  = 0xbf
+	bin8       = 0xc4
+	bin16      = 0xc5
+	bin32      = 0xc6
+	uint8tag   = 0xcc
+	uint16tag  = 0xcd
+	uint32tag  = 0xce
+	uint64tag  = 0xcf
+)
+
+// parser provides a zero-allocation parser for vote messages.
+type parser struct {
+	data []byte
+	pos  int
+}
+
+func newParser(data []byte) *parser {
+	return &parser{data: data}
+}
+
+// Error if we need more bytes than available
+func (p *parser) ensureBytes(n int) error {
+	if p.pos+n > len(p.data) {
+		return fmt.Errorf("unexpected EOF: need %d bytes, have %d", n, len(p.data)-p.pos)
+	}
+	return nil
+}
+
+// Read a single byte
+func (p *parser) readByte() (byte, error) {
+	if err := p.ensureBytes(1); err != nil {
+		return 0, err
+	}
+	b := p.data[p.pos]
+	p.pos++
+	return b, nil
+}
+
+// Read a fixmap header and return the count
+func (p *parser) readFixMap() (uint8, error) {
+	b, err := p.readByte()
+	if err != nil {
+		return 0, err
+	}
+
+	if b < fixMapMask || b > fixMapMax {
+		return 0, fmt.Errorf("expected fixmap, got 0x%02x", b)
+	}
+
+	return b & 0x0f, nil
+}
+
+// Zero-allocation string reading that returns a slice of the original data
+func (p *parser) readString() ([]byte, error) {
+	b, err := p.readByte()
+	if err != nil {
+		return nil, err
+	}
+	if b < fixStrMask || b > fixStrMax {
+		return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", b)
```

Comment:
> // Zero-allocation string reading that returns a slice of the original data

so format string makes things worse
```
network/vpack/msgp.go:87:69: b escapes to heap
```

```suggestion
	if b < fixStrMask || b > fixStrMax {
		val := b
		return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", val)
```
In this case `val` will escape only in case of an error making this function indeed zero allocs. Btw, the same applies to `readFixMap`, `readUintBytes` and `ensureBytes` where `n` escapes

Reply:
not really, when compiler determines a var escapes, its allocation goes to heap not stack/regs.

Reply:
I found some clues in x86 assembly
```
                return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", b)
  0xba4d0a              488d150ffe2500          LEAQ 0x25fe0f(IP), DX
  0xba4d11              4889542430              MOVQ DX, 0x30(SP)
  0xba4d16              0fb6d0                  MOVZX AL, DX
  0xba4d19              4c8d05e014b200          LEAQ runtime.staticuint64s(SB), R8
```
and
```
// staticuint64s is used to avoid allocating in convTx for small integer values.
// staticuint64s[0] == 0, staticuint64s[1] == 1, and so forth.
// It is defined in assembler code so that it is read-only.
var staticuint64s [256]uint64
```

so it appears to be some later optimization, LTO?